File: parser/projectParser.js
/**
 * projectParser.js
 * Orchestrates directory walking and delegates to language-specific parsers.
 * Builds the unified NodeGraph data structure.
 */

import { readdirSync, statSync, readFileSync } from 'fs';
import { join, relative, extname, basename } from 'path';
import { parseJavaScript } from './parsers/javascript.js';
import { parsePython } from './parsers/python.js';
import { parseJSON } from './parsers/json.js';
import { parseShell } from './parsers/shell.js';
import { resolveEdges } from './resolveEdges.js';
import { auditDocs } from './docCheck.js';
import { loadProjectConfig, initProjectConfig } from './projectConfig.js';

// Global ignore patterns — always applied regardless of project config
const GLOBAL_IGNORE = [
  // dependency / build artifacts
  'node_modules', '__pycache__', '.venv', 'venv',
  'dist', 'build', '.next', '.nuxt', 'coverage',
  '.cache', '.idea', '.vscode',
  // minified files
  '*.min.js', '*.bundle.js',
  // s-ide internals
  '.nodegraph.json', 'side.project.json',
  // version archives — snapshots, not live source
  'versions', 'releases', 'archive', 'archives',
  'VERSIONS', 'RELEASES', 'ARCHIVE',
  // git
  '.git',
  // uploads dir from s-ide itself
  'uploads',
];

const PARSERS = {
  '.js':   parseJavaScript,
  '.mjs':  parseJavaScript,
  '.cjs':  parseJavaScript,
  '.jsx':  parseJavaScript,
  '.ts':   parseJavaScript,
  '.tsx':  parseJavaScript,
  '.py':   parsePython,
  '.json': parseJSON,
  '.sh':   parseShell,
  '.bash': parseShell,
};

// File categories for visual grouping in the node editor
const FILE_CATEGORIES = {
  '.js': 'javascript', '.mjs': 'javascript', '.cjs': 'javascript',
  '.jsx': 'react', '.tsx': 'react',
  '.ts': 'typescript',
  '.py': 'python',
  '.json': 'config',
  '.sh': 'shell', '.bash': 'shell',
  '.css': 'style', '.scss': 'style', '.less': 'style',
  '.md': 'docs', '.mdx': 'docs',
  '.env': 'config', '.toml': 'config', '.yaml': 'config', '.yml': 'config',
  '.html': 'markup', '.htm': 'markup',
  '.sql': 'database',
};

function shouldIgnore(name, extraPatterns = []) {
  const patterns = [...GLOBAL_IGNORE, ...extraPatterns];
  return patterns.some(pattern => {
    if (pattern.startsWith('*')) {
      return name.endsWith(pattern.slice(1));
    }
    return name === pattern || name.startsWith('.');
  });
}

function walkDirectory(dir, rootDir, fileList = [], extraIgnore = []) {
  let entries;
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return fileList;
  }

  for (const entry of entries) {
    if (shouldIgnore(entry.name, extraIgnore)) continue;

    const fullPath = join(dir, entry.name);

    if (entry.isDirectory()) {
      walkDirectory(fullPath, rootDir, fileList, extraIgnore);
    } else if (entry.isFile()) {
      const ext = extname(entry.name).toLowerCase();
      fileList.push({
        fullPath,
        relativePath: relative(rootDir, fullPath),
        ext,
        name: entry.name,
        category: FILE_CATEGORIES[ext] || 'other',
      });
    }
  }

  return fileList;
}

function getFileStats(fullPath) {
  try {
    const stat = statSync(fullPath);
    return {
      size: stat.size,
      modified: stat.mtime.toISOString(),
    };
  } catch {
    return { size: 0, modified: null };
  }
}

function readFileSafe(fullPath) {
  try {
    return readFileSync(fullPath, 'utf-8');
  } catch {
    return null;
  }
}

function countLines(content) {
  return content ? content.split('\n').length : 0;
}

export async function parseProject(rootDir) {
  const startTime = Date.now();

  // Load per-project config (side.project.json) — graceful if missing
  // Load config, auto-creating side.project.json if absent
  const projectConfig = initProjectConfig(rootDir);
  const extraIgnore = projectConfig.ignore || [];
  if (extraIgnore.length) {
    // Also add the versions dir name from config in case it's non-standard
    const versionsDir = projectConfig.versions?.dir;
    if (versionsDir && !extraIgnore.includes(versionsDir)) extraIgnore.push(versionsDir);
  }

  const files = walkDirectory(rootDir, rootDir, [], extraIgnore);

  const nodes = [];
  const rawEdges = []; // edges before path resolution

  const fileIndex = new Map(); // relativePath -> nodeId

  // First pass: build all nodes
  for (const file of files) {
    const content = readFileSafe(file.fullPath);
    const stats = getFileStats(file.fullPath);
    const parser = PARSERS[file.ext];

    let parsed = {
      imports: [],
      exports: [],
      definitions: [],
      calls: [],
      errors: [],
    };

    if (parser && content !== null) {
      try {
        parsed = parser(content, file.fullPath);
      } catch (err) {
        parsed.errors.push(`Parse error: ${err.message}`);
      }
    }

    const nodeId = file.relativePath.replace(/[^a-zA-Z0-9]/g, '_');

    const node = {
      id: nodeId,
      label: file.name,
      path: file.relativePath,
      fullPath: file.fullPath,
      category: file.category,
      ext: file.ext,
      lines: countLines(content),
      size: stats.size,
      modified: stats.modified,
      // Semantic data
      imports: parsed.imports,
      exports: parsed.exports,
      definitions: parsed.definitions,
      calls: parsed.calls,
      errors: parsed.errors,
      // Layout hint for the node editor (set later by layout engine)
      position: null,
    };

    nodes.push(node);
    fileIndex.set(file.relativePath, nodeId);
  }

  // Second pass: resolve edges from raw imports
  const edges = resolveEdges(nodes, fileIndex, rootDir);

  // Collect language stats
  const langStats = {};
  for (const node of nodes) {
    const cat = node.category;
    if (!langStats[cat]) langStats[cat] = { files: 0, lines: 0 };
    langStats[cat].files++;
    langStats[cat].lines += node.lines;
  }

  // Auto-layout: assign positions using a layered approach
  assignPositions(nodes, edges);

  // Documentation health audit
  const docs = auditDocs(rootDir, nodes);

  return {
    version: '1.0.0',
    meta: {
      root: rootDir,
      parsedAt: new Date().toISOString(),
      parseTime: Date.now() - startTime,
      totalFiles: nodes.length,
      totalEdges: edges.length,
      languages: langStats,
      docs,
      project: {
        name:        projectConfig.name || rootDir.split('/').pop(),
        version:     projectConfig.version,
        description: projectConfig.description,
        run:         projectConfig.run,
        versions:    projectConfig.versions,
        hasConfig:   projectConfig._exists,
      },
    },
    nodes,
    edges,
  };
}

/**
 * Simple layered layout: group by category, spread nodes out
 * so the node editor has reasonable starting positions.
 * The user can rearrange in the UI afterward.
 */
function assignPositions(nodes, edges) {
  // Build a dependency depth map
  const depthMap = new Map();
  const nodeMap = new Map(nodes.map(n => [n.id, n]));
  
  // Build adjacency for depth calculation
  const incomingCount = new Map(nodes.map(n => [n.id, 0]));
  for (const edge of edges) {
    incomingCount.set(edge.target, (incomingCount.get(edge.target) || 0) + 1);
  }

  // Topological sort for layering
  const queue = nodes.filter(n => (incomingCount.get(n.id) || 0) === 0);
  const depths = new Map(nodes.map(n => [n.id, 0]));
  const visited = new Set();

  while (queue.length > 0) {
    const node = queue.shift();
    if (visited.has(node.id)) continue;
    visited.add(node.id);

    const nodeEdges = edges.filter(e => e.source === node.id);
    for (const edge of nodeEdges) {
      const newDepth = (depths.get(node.id) || 0) + 1;
      if (newDepth > (depths.get(edge.target) || 0)) {
        depths.set(edge.target, newDepth);
      }
      const targetNode = nodeMap.get(edge.target);
      if (targetNode) queue.push(targetNode);
    }
  }

  // Group nodes by depth
  const layers = new Map();
  for (const [id, depth] of depths) {
    if (!layers.has(depth)) layers.set(depth, []);
    layers.get(depth).push(id);
  }

  const NODE_W = 240;
  const NODE_H = 160;
  const LAYER_GAP = 320;
  const NODE_GAP = 200;

  for (const [depth, ids] of layers) {
    ids.forEach((id, i) => {
      const node = nodeMap.get(id);
      if (node) {
        node.position = {
          x: depth * LAYER_GAP,
          y: i * NODE_GAP - (ids.length * NODE_GAP) / 2,
        };
      }
    });
  }

  // Catch any unpositioned nodes (cycles, etc.)
  let orphanX = 0;
  for (const node of nodes) {
    if (!node.position) {
      node.position = { x: orphanX, y: -600 };
      orphanX += NODE_W + 40;
    }
  }
}

File: parser/project_parser.py
"""
parser/project_parser.py
=========================
Orchestrator: ties together the walker, per-language parsers, edge
resolver, layout engine, and doc auditor into a single parse pass.

Parse pipeline (with stage timings)
------------------------------------
1. init_project_config  — load/create side.project.json
2. walk_directory       — discover all source files
3. per-file parsing     — call the appropriate language parser
4. resolve_edges        — turn raw import strings into graph edges
5. assign_positions     — auto-layout for the node editor
6. audit_docs           — README / empty-module health check
7. write_graph_json     — auto-save .nodegraph.json to project root

Each stage is timed by a ParseTimer; results are stored in
graph.meta["perf"] for display in the GUI performance panel.

Auto-save
---------
After every successful parse, the full graph dict is written to:
    <project_root>/.nodegraph.json

This ensures the post-parse JSON always exists on disk without needing
the GUI to explicitly save it.
"""

from __future__ import annotations
import json
import os
from datetime import datetime, timezone

from parser.walker import walk_directory, make_node_id, FileInfo
from parser.parsers import PARSERS
import parser.parsers.python_parser   # ensure picklable
import concurrent.futures
from parser.resolve_edges import resolve_edges
from parser.layout import assign_positions
from parser.doc_check import audit_docs
from parser.project_config import init_project_config
from monitor.perf import ParseTimer

from graph.types import (
    FileNode, ImportRecord, ExportRecord, Definition,
    GraphMeta, ProjectGraph,
)


def _read_safe(path: str) -> str | None:
    try:
        with open(path, "r", encoding="utf-8", errors="replace") as f:
            return f.read()
    except Exception:
        return None


def _file_stats(path: str) -> tuple[int, str | None]:
    try:
        stat = os.stat(path)
        mtime = datetime.fromtimestamp(stat.st_mtime, tz=timezone.utc).isoformat()
        return stat.st_size, mtime
    except Exception:
        return 0, None


def _make_node(file_info, parsed: dict, size: int, mtime: str | None, content: str | None) -> FileNode:
    lines = content.count("\n") + 1 if content else 0

    def to_import(r):
        return r if isinstance(r, ImportRecord) else ImportRecord(**r)
    def to_export(r):
        return r if isinstance(r, ExportRecord) else ExportRecord(**r)
    def to_def(r):
        return r if isinstance(r, Definition) else Definition(**r)

    return FileNode(
        id=make_node_id(file_info.relative_path),
        label=file_info.name,
        path=file_info.relative_path,
        full_path=file_info.full_path,
        category=file_info.category,
        ext=file_info.ext,
        lines=lines,
        size=size,
        modified=mtime,
        imports=[to_import(r) for r in (parsed.get("imports") or [])],
        exports=[to_export(r) for r in (parsed.get("exports") or [])],
        definitions=[to_def(r) for r in (parsed.get("definitions") or [])],
        tags=list(parsed.get("tags") or []),
        errors=list(parsed.get("errors") or []),
    )
def _parse_file_worker(file_info: FileInfo) -> FileNode:
    """Worker task for ProcessPoolExecutor."""
    content = _read_safe(file_info.full_path)
    size, mtime = _file_stats(file_info.full_path)
    # Re-import parsers inside worker if needed, but they are top-level so should work
    from parser.parsers import PARSERS
    
    parser_fn = PARSERS.get(file_info.ext)
    if parser_fn and content is not None:
        try:
            parsed = parser_fn(content, file_info.full_path)
        except Exception as exc:
            parsed = {"imports": [], "exports": [], "definitions": [],
                      "tags": [], "errors": [f"Parser crash: {exc}"]}
    else:
        parsed = {"imports": [], "exports": [], "definitions": [],
                  "tags": [], "errors": []}
        
    return _make_node(file_info, parsed, size, mtime, content)


def parse_project(root_dir: str, save_json: bool = True) -> ProjectGraph:
    """
    Full parse pipeline for a project directory.

    Parameters
    ----------
    root_dir  : absolute or relative path to the project
    save_json : if True (default), write .nodegraph.json to root_dir

    Returns a ProjectGraph ready for JSON serialisation.
    Stage timings are available at graph.to_dict()["meta"]["perf"].
    """
    timer    = ParseTimer()
    root_dir = os.path.abspath(root_dir)

    # 1. Config
    with timer.stage("config"):
        config = init_project_config(root_dir)
        extra_ignore = list(config.get("ignore") or [])
        versions_dir = (config.get("versions") or {}).get("dir", "versions")
        if versions_dir and versions_dir not in extra_ignore:
            extra_ignore.append(versions_dir)
        # Also ignore logs/ and build output dirs
        for _d in ("logs", "dist", "build", ".nodegraph.json"):
            if _d not in extra_ignore:
                extra_ignore.append(_d)

    # 2. Walk
    with timer.stage("walk"):
        files = walk_directory(root_dir, extra_ignore=extra_ignore)

    # Load existing graph for delta-updates
    old_nodes: dict[str, dict] = {}
    if os.path.isfile(os.path.join(root_dir, ".nodegraph.json")):
        try:
            with open(os.path.join(root_dir, ".nodegraph.json"), "r") as f:
                old_data = json.load(f)
                for n in old_data.get("nodes", []):
                    old_nodes[n["path"]] = n
        except Exception:
            pass

    # 3. Parse files (Delta + Parallel)
    nodes: list[FileNode] = []
    file_index: dict[str, str] = {}
    changed_files: list[FileInfo] = []

    with timer.stage("parse_files"):
        for file_info in files:
            size, mtime = _file_stats(file_info.full_path)
            old = old_nodes.get(file_info.relative_path)
            
            # Re-use node if mtime and size match
            if old and old.get("modified") == mtime and old.get("size") == size:
                # Reconstruct FileNode from dict
                node = FileNode(
                    id=old["id"], label=old["label"], path=old["path"],
                    full_path=old.get("full_path", file_info.full_path),
                    category=old.get("category", "other"),
                    ext=old.get("ext", ".py"), lines=old.get("lines", 0),
                    size=size, modified=mtime,
                    imports=[ImportRecord(**r) for r in old.get("imports", [])],
                    exports=[ExportRecord(**r) for r in old.get("exports", [])],
                    definitions=[Definition(**r) for r in old.get("definitions", [])],
                    tags=list(old.get("tags") or []),
                    errors=list(old.get("errors") or []),
                )
                nodes.append(node)
                file_index[node.path] = node.id
            else:
                changed_files.append(file_info)

        if changed_files:
            # Parse only changed files in parallel
            with concurrent.futures.ProcessPoolExecutor() as executor:
                results = list(executor.map(_parse_file_worker, changed_files))
            for node in results:
                nodes.append(node)
                file_index[node.path] = node.id

    # 4. Edges
    with timer.stage("resolve_edges"):
        edges = resolve_edges(nodes, file_index, root_dir)

    # 5. Layout
    with timer.stage("layout"):
        assign_positions(nodes, edges)

    # 6. Docs
    with timer.stage("doc_audit"):
        docs = audit_docs(root_dir, nodes)

    # 7. Language stats
    lang_stats: dict[str, dict] = {}
    for node in nodes:
        cat = node.category
        if cat not in lang_stats:
            lang_stats[cat] = {"files": 0, "lines": 0}
        lang_stats[cat]["files"] += 1
        lang_stats[cat]["lines"] += node.lines

    # Finalise perf — include write_json stage if it runs
    # We create a preliminary report now; write_json will add its stage after
    perf = timer.report()

    meta = GraphMeta(
        root=root_dir,
        parsed_at=datetime.now(tz=timezone.utc).isoformat(),
        parse_time_ms=perf["total_ms"],
        total_files=len(nodes),
        total_edges=len(edges),
        languages=lang_stats,
        docs=docs,
        project_name=config.get("name") or os.path.basename(root_dir),
        project_version=config.get("version") or "0.0.0",
        project_description=config.get("description") or "",
        project_run=dict(config.get("run") or {}),
        has_config=bool(config.get("_exists")),
        perf=perf,
    )

    graph = ProjectGraph(version="1.0.0", meta=meta, nodes=nodes, edges=edges)

    # 8. Auto-save .nodegraph.json (timed; updates meta.perf after the fact)
    if save_json:
        with timer.stage("write_json"):
            _write_graph_json(graph, root_dir)
        # Refresh perf with the write_json stage included
        final_perf = timer.report()
        graph.meta = GraphMeta(
            **{k: v for k, v in vars(meta).items() if k != "perf"},
            perf=final_perf,
        )
        # Overwrite the JSON with the complete timing
        _write_graph_json(graph, root_dir)

    return graph


def _write_graph_json(graph: ProjectGraph, root_dir: str) -> str:
    """
    Serialise the graph to <root_dir>/.nodegraph.json.
    perf data is already in graph.meta.perf via the GraphMeta dataclass.
    Returns the path written.
    """
    out_path = os.path.join(root_dir, ".nodegraph.json")
    with open(out_path, "w", encoding="utf-8") as f:
        json.dump(graph.to_dict(), f, indent=2)
    return out_path
